### ISSUE TITLE
Not resolving parent class correctly

### ISSUE THREAD

**Issue by thomthom** [lsegal/yard#912]
I noticed that when generating docs for our API (written in C) doesn't resolve the parent class for all classes. Classes that subclass the same parent class doesn't consistently resolve.

I ran with `--debug` to try to figure out what is going on. Here's my preliminary finding:

Key baseclass `Sketchup::Entity` - Most of our classes inherit from this.

One class `Sketchup::Curve` resolve it's parent class `Sketchup::Entity` correctly.
![image](https://cloud.githubusercontent.com/assets/192418/11093014/e4a52d74-8887-11e5-8d06-082b242b3158.png)

But this other class `Sketchup::Drawingelement` fail:
![image](https://cloud.githubusercontent.com/assets/192418/11093029/f95f1fc2-8887-11e5-9d0e-9ca274332011.png)

It's a bit strange - as it display "Entity" as the parent name, but doesn't resolve to `Sketchup::Entity` which `Sketchup::Curve` does.

Excerpt from the debug log: (full filenames stripped)

```
# Line 90:
[debug]: Parsing [...]entity.cpp
[debug]: Missing object Sketchup in file `[...]entity.cpp', moving it to the back of the line.
# Line 595:
[debug]: Parsing [..]drawingelement.cpp
# Line 1446:
[debug]: Re-processing [..]entity.cpp...
```

So it looks like that YARD can push file back in the queue if the module it is defined it hasn't been parsed yet. But by the looks of it, YARD doesn't do the same for the base classes when creating subclasses.

In the above example it looks as though `drawingelement.cpp` should be pushed back in the queue until `entity.cpp` has been processed.

I might be wrong in my assessment - I'm not  familiar with exactly these things are resolved.

Any clues to what might be going on here - or how to further debug - would be much appreciated.
(I've not been able to figure out a minimal reproducible case yet.)

**Comment by thomthom** [comment]
Looking into another warning in the log:

```
[warn]: The proxy Material has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before [...]
```

Does this mean there is a way to specify the order of which files are processed?

**Comment by lsegal** [comment]
You can force files to parse in a specific order by providing the filename in your .yardopts file as the first file in the files glob argument:

```
file1.c *.c
```

**Comment by thomthom** [comment]
I managed to repo it with minimal code. GitHub doesn't allow ZIP uploads so I uploaded as a gist, you can either pull it or download the full set from there: https://gist.github.com/thomthom/7cd8d7954646fb5b1017

The example show a few files which defines classes that inherit from `Sketchup::Edge` - only `Sketchup::Layer` resolve its parent `Sketchup::Entity` class correctly in the docs.

The rest will just display `Entity` without a link back to its parent and without displaying any inherited methods.

On the surface of things it appear that it's because of loading order - where `Sketchup::Layer` resolves it base class correctly because it's parsed after.

**Comment by thomthom** [comment]
Adding a few explicit directives to control the order of some key classes and modules seemed to resolve it in my case. Thank you.

Though, as you say, it would be nice if it was automatically resolved such that only could simply provide a wild-card match for all CPP files.

**Comment by lsegal** [comment]
@thomthom are you suggesting that adding .cpp to the glob would resolve this for you? That seems like an easy enough change.

**Comment by thomthom** [comment]
Using a wild-card in the `yarddoc` argument was what I used (`**/*cpp`), but the problem was that the files parsed where done in the order the `glob` method would find them. Yard manage to bump files back in the queue when the namespace they belong to haven't been parsed. But Yard doesn't bump the files down if a class inherit from another class which is yet to be parsed. This is where I had to manually list the files containing base classes first - before my wild-card pattern.

I don't think that commit will resolve this issue I'm afraid.

**Comment by lsegal** [comment]
> files parsed where done in the order the glob method would find them.

Files are not parsed in glob-order, they are sorted by length prior to parsing. That said, the issue you're referring to should be solvable by adding `ensure_loaded!(parent_class)` guard prior to [the proxy check in handler_methods.rb](https://github.com/lsegal/yard/blob/master/lib/yard/handlers/c/handler_methods.rb#L19). I can take a look at that one, but I haven't been able to reproduce your unlinked Entity issue from the screenshots on my machine with master... are you sure that your gist reliably reproduces the issue?

**Comment by thomthom** [comment]
On my machine, with latest on `master` the `Edge` and `Layer` class both produce unlinked reference to `Entity`:

![2015-12-05_16h26_55](https://cloud.githubusercontent.com/assets/192418/11608578/732a46ae-9b6d-11e5-9e1f-a5381122ac4f.png)

I'm on a Windows system - not sure if that has any significance?

I added `ensure_loaded!(parent_class)` to line 19 before the `Proxy` check - and that did not seem to work. The docs generated now said that `Sketchup::Layer` inherited `Object` directly.

The debug log:

```
C:\Users\Thomas\Desktop\temp
λ yardoc *.cpp --debug
[debug]: Parsing ["attributes.cpp", "drawingelement.cpp", "edge.cpp", "entity.cpp", "layer.cpp", "sketchup.cpp"] with `ruby` parser
[debug]: Parsing attributes.cpp
[debug]: Missing object Sketchup in file `attributes.cpp', moving it to the back of the line.
[debug]: Parsing drawingelement.cpp
[debug]: Missing object Sketchup in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Parsing edge.cpp
[debug]: Missing object Sketchup in file `edge.cpp', moving it to the back of the line.
[debug]: Parsing entity.cpp
[debug]: Missing object Sketchup in file `entity.cpp', moving it to the back of the line.
[debug]: Parsing layer.cpp
[debug]: Missing object Sketchup in file `layer.cpp', moving it to the back of the line.
[debug]: Parsing sketchup.cpp
[debug]: Re-processing layer.cpp...
[debug]: Missing object Entity in file `layer.cpp', moving it to the back of the line.
[debug]: Re-processing layer.cpp...
[debug]: Missing object Entity in file `layer.cpp', moving it to the back of the line.
[debug]: Re-processing layer.cpp...
[warn]: The proxy Entity has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before `layer.cpp'
[warn]:
[debug]: Missing object Layer in file `layer.cpp', moving it to the back of the line.
[debug]: Re-processing layer.cpp...
[debug]: Missing object Layer in file `layer.cpp', moving it to the back of the line.
[debug]: Re-processing layer.cpp...
[debug]: Re-processing entity.cpp...
[debug]: Missing object Object in file `entity.cpp', moving it to the back of the line.
[debug]: Re-processing entity.cpp...
[debug]: Missing object Object in file `entity.cpp', moving it to the back of the line.
[debug]: Re-processing entity.cpp...
[warn]: The proxy Object has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before `entity.cpp'
[warn]:
[debug]: Missing object Entity in file `entity.cpp', moving it to the back of the line.
[debug]: Re-processing entity.cpp...
[debug]: Missing object Entity in file `entity.cpp', moving it to the back of the line.
[debug]: Re-processing entity.cpp...
[debug]: Re-processing edge.cpp...
[debug]: Missing object Drawingelement in file `edge.cpp', moving it to the back of the line.
[debug]: Re-processing edge.cpp...
[debug]: Missing object Drawingelement in file `edge.cpp', moving it to the back of the line.
[debug]: Re-processing edge.cpp...
[warn]: The proxy Drawingelement has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before `edge.cpp'
[warn]:
[debug]: Missing object Edge in file `edge.cpp', moving it to the back of the line.
[debug]: Re-processing edge.cpp...
[debug]: Missing object Edge in file `edge.cpp', moving it to the back of the line.
[debug]: Re-processing edge.cpp...
[debug]: Re-processing drawingelement.cpp...
[debug]: Missing object Entity in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Re-processing drawingelement.cpp...
[debug]: Missing object Entity in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Re-processing drawingelement.cpp...
[warn]: The proxy Entity has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before `drawingelement.cpp'
[warn]:
[debug]: Missing object Drawingelement in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Re-processing drawingelement.cpp...
[debug]: Missing object Drawingelement in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Re-processing drawingelement.cpp...
[debug]: Re-processing attributes.cpp...
[debug]: Missing object Entity in file `attributes.cpp', moving it to the back of the line.
[debug]: Re-processing attributes.cpp...
[debug]: Missing object Entity in file `attributes.cpp', moving it to the back of the line.
[debug]: Re-processing attributes.cpp...
[warn]: The proxy Entity has not yet been recognized.
[warn]: If this class/method is part of your source tree, this will affect your documentation results.
[warn]: You can correct this issue by loading the source file for this object before `attributes.cpp'
[warn]:
[debug]: Missing object Attribs in file `attributes.cpp', moving it to the back of the line.
[debug]: Re-processing attributes.cpp...
[debug]: Missing object Attribs in file `attributes.cpp', moving it to the back of the line.
[debug]: Re-processing attributes.cpp...
[debug]: Serializing to .yardoc/objects/root.dat
[debug]: Re-generating object ...
[debug]: Re-generating object Sketchup...
[debug]: Re-generating object Sketchup::AttributesDictionary...
[debug]: Re-generating object Sketchup::Drawingelement...
[debug]: Re-generating object Sketchup::Edge...
[debug]: Re-generating object Sketchup::Entity...
[debug]: Re-generating object Sketchup::Layer...
[debug]: Generating asset js/jquery.js
[debug]: Serializing to doc/js/jquery.js
[debug]: Generating asset js/app.js
[debug]: Serializing to doc/js/app.js
[debug]: Generating asset js/full_list.js
[debug]: Serializing to doc/js/full_list.js
[debug]: Generating asset css/style.css
[debug]: Serializing to doc/css/style.css
[debug]: Generating asset css/common.css
[debug]: Serializing to doc/css/common.css
[debug]: Generating asset css/full_list.css
[debug]: Serializing to doc/css/full_list.css
[debug]: Generating asset class_list.html
[debug]: Serializing to doc/class_list.html
[debug]: Generating asset method_list.html
[debug]: Serializing to doc/method_list.html
[debug]: Generating asset file_list.html
[debug]: Serializing to doc/file_list.html
[debug]: Generating asset frames.html
[debug]: Serializing to doc/frames.html
[debug]: Serializing to doc/index.html
[debug]: Serializing to doc/_index.html
[debug]: Serializing to doc/top-level-namespace.html
[debug]: Serializing to doc/Sketchup.html
[debug]: Serializing to doc/Sketchup/AttributesDictionary.html
[debug]: Serializing to doc/Sketchup/Drawingelement.html
[debug]: Serializing to doc/Sketchup/Edge.html
[debug]: Serializing to doc/Sketchup/Entity.html
[debug]: Serializing to doc/Sketchup/Layer.html
Files:           2
Modules:         1 (    1 undocumented)
Classes:         5 (    5 undocumented)
Constants:       0 (    0 undocumented)
Methods:         6 (    6 undocumented)
 0.00% documented
```

It looks like it get stuck in a loop. Once it processes `sketchup.cpp` it tried to process `layer.cpp` where it then misses `Entity` - so it puts `layer.cpp` back in the queue only to pick it right up again.

When I go back to the unmodified trunk the debug log looks like this:

```
C:\Users\Thomas\Desktop\temp
λ yardoc *.cpp --debug
[debug]: Parsing ["attributes.cpp", "drawingelement.cpp", "edge.cpp", "entity.cpp", "layer.cpp", "sketchup.cpp"] with `ruby` parser
[debug]: Parsing attributes.cpp
[debug]: Missing object Sketchup in file `attributes.cpp', moving it to the back of the line.
[debug]: Parsing drawingelement.cpp
[debug]: Missing object Sketchup in file `drawingelement.cpp', moving it to the back of the line.
[debug]: Parsing edge.cpp
[debug]: Missing object Sketchup in file `edge.cpp', moving it to the back of the line.
[debug]: Parsing entity.cpp
[debug]: Missing object Sketchup in file `entity.cpp', moving it to the back of the line.
[debug]: Parsing layer.cpp
[debug]: Missing object Sketchup in file `layer.cpp', moving it to the back of the line.
[debug]: Parsing sketchup.cpp
[debug]: Re-processing layer.cpp...
[debug]: Re-processing entity.cpp...
[debug]: Re-processing edge.cpp...
[debug]: Re-processing drawingelement.cpp...
[debug]: Re-processing attributes.cpp...
[debug]: Serializing to .yardoc/objects/root.dat
[debug]: Re-generating object ...
[debug]: Re-generating object Sketchup...
[debug]: Re-generating object Sketchup::AttributesDictionary...
[debug]: Re-generating object Sketchup::Drawingelement...
[debug]: Re-generating object Sketchup::Edge...
[debug]: Re-generating object Sketchup::Entity...
[debug]: Re-generating object Sketchup::Layer...
[debug]: Generating asset js/jquery.js
[debug]: Serializing to doc/js/jquery.js
[debug]: Generating asset js/app.js
[debug]: Serializing to doc/js/app.js
[debug]: Generating asset js/full_list.js
[debug]: Serializing to doc/js/full_list.js
[debug]: Generating asset css/style.css
[debug]: Serializing to doc/css/style.css
[debug]: Generating asset css/common.css
[debug]: Serializing to doc/css/common.css
[debug]: Generating asset css/full_list.css
[debug]: Serializing to doc/css/full_list.css
[debug]: Generating asset class_list.html
[debug]: Serializing to doc/class_list.html
[debug]: Generating asset method_list.html
[debug]: Serializing to doc/method_list.html
[debug]: Generating asset file_list.html
[debug]: Serializing to doc/file_list.html
[debug]: Generating asset frames.html
[debug]: Serializing to doc/frames.html
[debug]: Serializing to doc/index.html
[debug]: Serializing to doc/_index.html
[debug]: Serializing to doc/top-level-namespace.html
[debug]: Serializing to doc/Sketchup.html
[debug]: Serializing to doc/Sketchup/AttributesDictionary.html
[debug]: Serializing to doc/Sketchup/Drawingelement.html
[debug]: Serializing to doc/Sketchup/Edge.html
[debug]: Serializing to doc/Sketchup/Entity.html
[debug]: Serializing to doc/Sketchup/Layer.html
Files:           7
Modules:         1 (    1 undocumented)
Classes:         5 (    5 undocumented)
Constants:       0 (    0 undocumented)
Methods:         6 (    6 undocumented)
 0.00% documented
```

First line says: `[debug]: Parsing ["attributes.cpp", "drawingelement.cpp", "edge.cpp", "entity.cpp", "layer.cpp", "sketchup.cpp"] with `ruby` parser`
All of these files need the `Sketchup` module defined in `sketchup.cpp` - so they are queued.
Last file to be pushed back into the queue is `layer.cpp` - then `sketchup.cpp` is processed and it resumes the queue. The next item in the queue appear to be the last one in - looks like it's a stack pattern being used.
However, when you look at what happens with `ensure_loaded!(parent_class)` it keeps pushing `layers.cpp` back in the stack, only to pick it right up again without trying to process the other onces first.

You mentioned that you don't see the same "unlinked" issue with my gist as I do, What does your debug log look like?

**Comment by lsegal** [comment]
> You mentioned that you don't see the same "unlinked" issue with my gist as I do, What does your debug log look like?

I see exactly what you pasted in your second log (unmodified source). 

Just to clarify, I cannot reproduce your 1st screenshot, but I *can* repro your second (Sketchup::Edge does not link Drawingelement).

I took another look at resolving this. Using `ensure_loaded!()` doesn't solve it, there's a similar but not quite-the-same lookup that has to happen. Unfortunately YARD has a limitation in that it can only be searching for 1 single "missing object" at a time, so adding this check wouldn't work without a re-architecting of the parser. The problem is that Layer needs Entity, and Entity needs Sketchup-- by the time Layer is reparsed, Sketchup has been added, but Entity hasn't yet been. This would be solved if you parsed sketchup.cpp first, that way the rest of your files would be missing Entity, but *only* Entity, and that would be quickly solved by a single "re-parse" of the other files. Indeed, adding `ext/sketchup.cpp ext/**/*.cpp` as my files list solves all linking issues for me.

FWIW this is one reason why YARD sorts its list of files to parse by length. If you follow typical Ruby code structure conventions, objects inside of a namespace sit inside a directory named after said namespace, whereas the namespace (module / class) definition itself sits in the parent directory. In your case, the structure would be:

```
ext/
`- sketchup.cpp
`- sketchup/
   `- entity.cpp
   `- layer.cpp
   `- ...
```

YARD would parse the above by selecting `ext/sketchup.cpp` first. That's why this kind of lookup failure doesn't often happen with Ruby code. Using this structure would solve many of the code-dependency issues that YARD is running into. I would recommend using this organization since it's consistent with Ruby's organization conventions and things will "just work".

The tl;dr here is that this specific issue is a limitation in YARD and not "fixable" (technically possible but is too complex to be worth it)-- but you can use either of the above workarounds.

**Comment by thomthom** [comment]
Thanks for looking into that.
Yea, we don't have standard structure to our Ruby code, ~10 code base, hundreds of classes/modules - I think we'll stick with simply manually ordering the important base classes and root namespaces in our yardopts file. I'll be a one time setup anyway and it works.

I'd hoped this would be an issue where I'd be able to contribute back, but alas. I'll see if I can take a stab at another one. :)

I'm ok with closing this one - unless you want to keep it open?

**Comment by lsegal** [comment]
There is still value in pulling in the above ensure loaded style fix (though not exactly that line) to warn about invalid usage. It would make it more clear when your codebase has these dependency issues (though it won't solve the issue for you). I'll make that change and then close this out.

**Comment by lsegal** [comment]
> I'd hoped this would be an issue where I'd be able to contribute back, but alas. I'll see if I can take a stab at another one. :)

FWIW there is still #915 which seems like something that can be fairly easily implemented, if you're looking to contribute.

**Comment by thomthom** [comment]
Yes, that was on my list of issues to look into.

Btw, when I ran with the `ensure_loaded!` added earlier I also got errors then - as it got stuck trying to resolve `Object` for some reason (`rb_cObject`). Did you get that when running my gist?
(Had added `sketchup.cpp` before `*.cpp`.)

**Comment by lsegal** [comment]
Using `ensure_loaded!` will cause the above error, yes. What actually has to be checked for is not the definition of the object, but rather, the definition of the variable. In other words, not `parent_class`, but the `parent` variable name itself has to be resolved. There also needs to be a check to avoid any core classes, since those would never get resolved.

**Comment by lsegal** [comment]
Finally had a chance to take a look at this and fix. YARD (master) should now attempt to look at your other files to resolve a namespace variable instead of just taking the literal name as a proxy.

**Comment by thomthom** [comment]
Oh, thank you very much for looking into that. Sorry for being slow at following up on the issues I said I'd look into. The documentation project got suspended for a bit due to some other urgent project. I'm scheduled to resume soon though.
I will do a quick test of the change you made tomorrow to verify though.
